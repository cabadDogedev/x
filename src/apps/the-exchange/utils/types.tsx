import { BridgingProvider, ExchangeOffer } from '@etherspot/prime-sdk/dist/sdk/data';
import { Route } from '@lifi/sdk';

export enum CardPosition {
    SWAP = 'SWAP',
    RECEIVE = 'RECEIVE'
}

export type SwapType = {
    fromAmount: number,
    fromTokenAddress: string,
    fromChainId: number;
    fromTokenDecimals: number;
    toTokenAddress: string,
    toChainId: number,
    toTokenDecimals: number;
    slippage: number;
    fromAccountAddress?: string
    provider?: BridgingProvider;
}

export type SwapOffer = {
    tokenAmountToReceive: number;
    offer: Route | ExchangeOffer;
}

export type AmountType = {
    tokenAmount: number;
    usdAmount: number;
  }
  
export type ChainType = {
    chainId: number;
    chainName: string;
  }
