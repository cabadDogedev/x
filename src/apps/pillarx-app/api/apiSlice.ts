import { createApi, fetchBaseQuery } from '@reduxjs/toolkit/query/react'
import { ApiResponse } from '../../../types/api'

export const pillarXApi = createApi({
    reducerPath: 'pillarXApi',
    baseQuery: fetchBaseQuery({ baseUrl: 'https://feed-nubpgwxpiq-uc.a.run.app/' }),
    endpoints: (builder) => ({
        getTilesInfo: builder.query<ApiResponse, string>({
            query: () => ''
        })
    })
})


export const { useGetTilesInfoQuery } = pillarXApi;
