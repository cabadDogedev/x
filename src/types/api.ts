export type Asset = {
    name: string;
    symbol: string;
    id: number;
    contracts: string[];
    logo: string;
    blockchains: string[];
  }
  
  export type CrossChainBalance = {
    balance: number;
    balanceRaw: string;
    chainId: string;
    address: string;
  }
  
  export type ContractBalance = {
    balance: number;
    balanceRaw: string;
    chainId: string;
    address: string;
    decimals: number;
  }
  
  export type AssetData = {
    asset: Asset;
    realized_pnl: number;
    unrealized_pnl: number;
    allocation: number;
    price: number;
    price_bought: number;
    price_change_24h: number;
    price_change_1h: number;
    total_invested: number;
    min_buy_price: number;
    max_buy_price: number;
    estimated_balance: number;
    token_balance: number;
    cross_chain_balances: {
      [key: string]: CrossChainBalance;
    };
    contracts_balances: ContractBalance[];
  }
  
  export type TotalPnlHistory = {
    realized: number;
    unrealized: number;
  }
  
  export type ApiData = {
    total_wallet_balance: number;
    wallet: string;
    wallets: string[];
    total_realized_pnl: number;
    total_unrealized_pnl: number;
    assets: AssetData[];
    total_pnl_history: {
      [key in '24h' | '7d' | '30d' | '1y']?: TotalPnlHistory;
  };
  
  }
  
  // TO DO - meta type to change when api ready, layout should be enum
  export type Projection = {
    meta: unknown;
    data: ApiData;
    layout: string;
    id: string;
  }
  
  export type ApiResponse = {
    projection: Projection[];
  }
  