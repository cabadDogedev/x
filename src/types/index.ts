export interface AppManifest {
  title: string;
  description: string;
  translations: Record<string, string>;
}
