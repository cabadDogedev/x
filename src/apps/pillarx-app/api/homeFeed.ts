/* eslint-disable no-console */
import { createApi, fetchBaseQuery } from '@reduxjs/toolkit/query/react';
import { addMiddleware } from '../../../store';
import { ApiResponse } from '../../../types/api';

export const homeFeedApi = createApi({
    reducerPath: 'homeFeedApi',
    baseQuery: fetchBaseQuery({ baseUrl:
        process.env.REACT_APP_USE_TESTNETS === 'true' ?
        'https://feed-nubpgwxpiq-uc.a.run.app' :
        'https://feed-7eu4izffpa-uc.a.run.app'
    }),
    endpoints: (builder) => ({
        getTilesInfo: builder.query<ApiResponse, { page: number; address: string }>({
            query: ({ page, address }) => `?page=${page}&address=${address}`,
        }),
    })
})

/**
 * Add this to the store
 */
addMiddleware(homeFeedApi);

export const { useGetTilesInfoQuery } = homeFeedApi;
