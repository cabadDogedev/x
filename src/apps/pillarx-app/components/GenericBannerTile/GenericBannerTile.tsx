import { useState } from 'react';

// types
import { Projection } from '../../../../types/api';

// components
import TileContainer from '../TileContainer/TileContainer'
import SkeletonLoader from '../../../../components/SkeletonLoader';
import Body from '../Typography/Body';
import H1 from '../Typography/H1';
import BodySmall from '../Typography/BodySmall';

type GenericBannerTileProps = {
    data: Projection | undefined;
    isDataLoading: boolean;
}

const GenericBannerTile = ({ data, isDataLoading }: GenericBannerTileProps) => {
    const { meta } = data || {};
    const [errorLink, setErrorLink] = useState<string | undefined>(undefined);

    const handleOpenBannerLink = () => {
        if (meta?.display.cta?.href) {
            window.open(meta.display.cta.href, '_blank');
        } else {
            setErrorLink('Oops, the link is not available');
            setTimeout(() => setErrorLink(undefined), 5000);
        }
    };

    if (!data || isDataLoading) {
        return (
        <TileContainer className='p-1 flex-col'>
            <div className='mt-32 p-10 tablet:mt-16 mobile:mt-9 mobile:p-4' data-testid='generic-banner-loading'>
                <SkeletonLoader  $height='30px' $width='180px' $marginBottom='10px' $radius='6px' />
                <SkeletonLoader  $height='20px' $width='150px' $marginBottom='16px' $radius='6px' />
                <SkeletonLoader  $height='35px' $width='100px' $marginBottom='16px' $radius='6px' />
            </div>
        </TileContainer>
        )
    }

    return (
        <TileContainer className={`p-1 flex-col ${!meta?.display && 'hidden'}`}>
            <div className="flex flex-col rounded-2xl  bg-cover bg-no-repeat bg-center" style={{ backgroundImage: meta?.display.backgroundImage && `url(${meta.display.backgroundImage})` }}>
                <div className='mt-32 p-10 tablet:mt-16 mobile:mt-9 mobile:p-4'>
                    <H1 className='text-[45px] tablet:leading-[67.5px] desktop:leading-[67.5px] mobile:text-xl mobile:leading-[30px]'>{meta?.display.title && meta.display.title}</H1>
                    <Body className='text-[22px] tablet:leading-[33px] desktop:leading-[33px] mobile:text-sm'>{meta?.display.subtitle && meta.display.subtitle}</Body>
                    <button className='mt-10 mobile:mt-4' onClick={handleOpenBannerLink}>
                        <Body className='bg-container_grey rounded-md py-3 px-5 mobile:text-sm mobile:py-2 mobile:px-4'>{meta?.display.cta?.text && meta.display.cta.text}</Body>
                    </button>
                    {errorLink && <BodySmall>{errorLink}</BodySmall>}
                </div>
            </div>
        </TileContainer>
    )
};

export default GenericBannerTile;
