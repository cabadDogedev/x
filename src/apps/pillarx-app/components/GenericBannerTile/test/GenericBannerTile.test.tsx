import renderer from 'react-test-renderer';
import { render, screen, fireEvent } from '@testing-library/react';

// types
import { ApiLayout, Projection } from '../../../../../types/api';

// components
import GenericBannerTile from '../GenericBannerTile';

describe('<GenericBannerTile />', () => {
    const mockDataGenericBanner: Projection = {
        meta: {
            display: {
                title: 'Banner title',
                subtitle: 'Banner subtitle',
                backgroundImage: 'https://example.com/background-image.png',
                cta: {
                    text: 'This is the button text',
                    href: 'https://example.com',
                },
            },
        },
        data: undefined,
        layout: ApiLayout.GENERIC_BANNER,
        id: 'genericBanner'
    };


    const mockDataGenericBannerWithoutCTA = {
        meta: {
            display: {
                title: 'Banner title',
                subtitle: 'Banner subtitle',
                backgroundImage: 'https://example.com/background-image.png',
                cta: {
                    text: 'This is the button text',
                    href: undefined,
                },
            },
        },
        data: undefined,
        layout: ApiLayout.GENERIC_BANNER,
        id: 'genericBanner'
    };

    it('renders correctly and matches snapshot', () => {
        const tree = renderer.create(
            <GenericBannerTile data={mockDataGenericBanner} isDataLoading={false} />
        ).toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('renders skeleton loader when data is loading', () => {
        render(<GenericBannerTile data={undefined} isDataLoading={true} />);

        expect(screen.getByTestId('generic-banner-loading')).toBeInTheDocument();
    });

    it('renders correctly with all the given data', () => {
        render(<GenericBannerTile data={mockDataGenericBanner} isDataLoading={false} />);

        expect(screen.getByText('Banner title')).toBeInTheDocument();
        expect(screen.getByText('Banner subtitle')).toBeInTheDocument();
        expect(screen.getByText('This is the button text')).toBeInTheDocument();
        expect(screen.getByRole('button')).toBeInTheDocument();
    });


    it('calls window.open when button is clicked', () => {
        window.open = jest.fn();

        render(<GenericBannerTile data={mockDataGenericBanner} isDataLoading={false} />);
        fireEvent.click(screen.getByText('This is the button text'));

        expect(window.open).toHaveBeenCalledWith('https://example.com', '_blank');
    });

    it('displays error message when button is clicked and href is missing', () => {
        render(<GenericBannerTile data={mockDataGenericBannerWithoutCTA} isDataLoading={false} />);
        fireEvent.click(screen.getByText('This is the button text'));

        expect(screen.getByText('Oops, the link is not available')).toBeInTheDocument();
    });
});
