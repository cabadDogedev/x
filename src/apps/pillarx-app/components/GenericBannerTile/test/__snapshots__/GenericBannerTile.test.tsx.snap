// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<GenericBannerTile /> renders correctly and matches snapshot 1`] = `
<div
  className="flex bg-container_grey rounded-2xl p-1 flex-col false"
>
  <div
    className="flex flex-col rounded-2xl  bg-cover bg-no-repeat bg-center"
    style={
      {
        "backgroundImage": "url(https://example.com/background-image.png)",
      }
    }
  >
    <div
      className="mt-32 p-10 tablet:mt-16 mobile:mt-9 mobile:p-4"
    >
      <h1
        className="text-3xl leading-[45px] font-medium mobile:text-xl mobile:leading-5 text-[45px] tablet:leading-[67.5px] desktop:leading-[67.5px] mobile:text-xl mobile:leading-[30px]"
      >
        Banner title
      </h1>
      <p
        className="text-base font-medium text-[22px] tablet:leading-[33px] desktop:leading-[33px] mobile:text-sm"
      >
        Banner subtitle
      </p>
      <button
        className="mt-10 mobile:mt-4"
        onClick={[Function]}
      >
        <p
          className="text-base font-medium bg-container_grey rounded-md py-3 px-5 mobile:text-sm mobile:py-2 mobile:px-4"
        >
          This is the button text
        </p>
      </button>
    </div>
  </div>
</div>
`;
